search.py

```diff
@@ -1,36 +1,49 @@
-    
-    #(I have excluded the outer two diagonals (on either side) of Pascal's triangle, because we know that the numbers in those diagonals (across the whole triangle) are an infinite number of 1s and two appearances of all other positive integers apart from 2 (for which there is only one appearance).)
-    
-    for n in range(4,row_limit):
-        for r in range(2,n-1):
-            nCr = math.factorial(n)/(math.factorial(n-r)*math.factorial(r))
-            list_numbers.append(nCr)
-    
-    set_numbers = set(list_numbers)
-    how_many_times = []
-    which_ones = []
-    
-    #Below I am checking how many times each distinct number appears in the cells we've generated. If a number appears more than twice (therefore more than four times across the whole triangle) I am storing both the number and the minimum number of times it appears across the whole triangle.
-    
-    for number in set_numbers:
-        count = list_numbers.count(number)
-        if count >= 3:
-            how_many_times.append(count+2)
-            which_ones.append(number)
-        
-    final_info = dict(zip(which_ones,how_many_times))
-    
-    for entry,frequency in final_info.items(): 
-        print ("The number {0} appears at least {1} times across the whole triangle".format(entry,frequency))
-        
-    return "END"
-
-search(501)
+# Use this code to find which numbers repeat an interesting number of times in Pascal's triangle.
+
+import time
+from collections import defaultdict
+
+def search(row_limit: int, minimum_counts: int = 5) -> [(int, int)]:
+    # A dictionary representing how many times a number appears in the cells we've generated. Every number appears once in the outer diagonals,
+    # so we can begin the counts at 2
+    counts = defaultdict(lambda: 2)
+
+    # Below I am dictating the rows and columns I will compute cell values for.
+
+    # (I have excluded the outer two diagonals (on either side) of Pascal's triangle, because we know that the numbers in those diagonals (across the whole triangle) are an infinite number of 1s and two appearances of all other positive integers apart from 2 (for which there is only one appearance).)
+
+    # Faster method for calculating subsequent nCr for performance reasons
+    # While this part was by no means the bottleneck, it's still nice to have
+    for n in range(4, row_limit):
+        nCr = n
+        for r in range(2, n - 1):
+            nCr *= n - r + 1
+            nCr //= r
+            counts[nCr] += 1
+
+    # Below I am checking how many times each distinct number appears in the cells we've generated. If a number appears more than four times, print it.
+
+    matches = list()
+
+    for entry, frequency in counts.items():
+        if frequency < minimum_counts:
+            continue
+
+        matches.append((entry, frequency))
+
+    return matches
+
+
+if __name__ == '__main__':
+    start_time = time.time()
+
+    matches = search(3001, 5)
```

Comment:
runs in ~5 seconds on my computer, I can change this to anything. also the '5' is optional, you can just say: `search(3001)`

Reply:
You can use parts of the standard library to reduce the work, namely `collections` and `itertools`, e.g.:

    import itertools as it
    from collections import Counter

    def search(limit):
        c = Counter(it.chain.from_iterable(it.accumulate(range(3, n - 1),
                                                         lambda x, r: x * (n - r + 1) // r,
                                                         initial=n * (n - 1) // 2) for n in range(4, limit)))
        return it.takewhile(lambda i: i[1] >= 4, c.most_common())

    for entry, frequency in search(3000):
        print(f"The number {entry} appears at least {frequency+2} times across the whole triangle")

Reply:
The `math.factorial` code was by far the biggest consumer of resources.

Reply:
Hey @achmpion, thanks for your review. As I mentioned in my (admittedly long) description above, the factorial code is not the bottleneck, the O(n²) brought by `list.count` is. Performance gains by improving the factorial code (by perhaps using `math.comb` or my above smart combinatorics code) are negligible before removing `list.count`.

So I think your comment is a bit misleading.

I would rephrase: The `math.factorial()` code was the biggest consumer of resources after `list.count()` to something less misleading, like: `list.count()` is by far the bottleneck and biggest consumer of resources, after which `math.factorial()` starts to become non-negligible.